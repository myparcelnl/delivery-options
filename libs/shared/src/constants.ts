import {PlatformName} from '@myparcel/constants';

export enum SubscriptionType {
  Required = 1,
  Never = 0,
  Optional = -1,
}

export const DEFAULT_PLATFORM = PlatformName.MyParcel;

export const FORM_NAME_DELIVERY_OPTIONS = 'deliveryOptions';

export enum CarrierSetting {
  AllowDeliveryOptions = 'allowDeliveryOptions',
  AllowEveningDelivery = 'allowEveningDelivery',
  AllowMondayDelivery = 'allowMondayDelivery',
  AllowMorningDelivery = 'allowMorningDelivery',
  AllowOnlyRecipient = 'allowOnlyRecipient',
  AllowPackageTypeDigitalStamp = 'allowPackageTypeDigitalStamp',
  AllowPackageTypeMailbox = 'allowPackageTypeMailbox',
  AllowPickupLocations = 'allowPickupLocations',
  AllowSameDayDelivery = 'allowSameDayDelivery',
  AllowSaturdayDelivery = 'allowSaturdayDelivery',
  /** @deprecated use ShowDeliveryDate instead */
  AllowShowDeliveryDate = 'allowShowDeliveryDate',
  AllowSignature = 'allowSignature',
  CutoffTime = 'cutoffTime',
  CutoffTimeSameDay = 'cutoffTimeSameDay',
  DeliveryDaysWindow = 'deliveryDaysWindow',
  DropOffDays = 'dropOffDays',
  DropOffDelay = 'dropOffDelay',
  FridayCutoffTime = 'fridayCutoffTime',
  PackageType = 'packageType',
  PriceEveningDelivery = 'priceEveningDelivery',
  PriceMondayDelivery = 'priceMondayDelivery',
  PriceMorningDelivery = 'priceMorningDelivery',
  PriceOnlyRecipient = 'priceOnlyRecipient',
  PricePackageTypeDigitalStamp = 'pricePackageTypeDigitalStamp',
  PricePackageTypeMailbox = 'pricePackageTypeMailbox',
  PricePickup = 'pricePickup',
  PriceSameDayDelivery = 'priceSameDayDelivery',
  PriceSaturdayDelivery = 'priceSaturdayDelivery',
  PriceSignature = 'priceSignature',
  PriceStandardDelivery = 'priceStandardDelivery',
  SaturdayCutoffTime = 'saturdayCutoffTime',
  ShowDeliveryDate = 'showDeliveryDate',
}

export enum ConfigSetting {
  ApiBaseUrl = 'apiBaseUrl',
  CarrierSettings = 'carrierSettings',
  Currency = 'currency',
  Locale = 'locale',
  PickupLocationsDefaultView = 'pickupLocationsDefaultView',
  PickupLocationsMapTileLayerData = 'pickupLocationsMapTileLayerData',
  PickupShowDistance = 'pickupShowDistance',
  Platform = 'platform',
  ShowDeliveryDate = 'showDeliveryDate',
  ShowPriceSurcharge = 'showPriceSurcharge',
  ShowPrices = 'showPrices',
}

export enum PickupLocationsView {
  Map = 'map',
  List = 'list',
}

export enum ComponentName {
  Button = 'Button',
  CheckboxGroup = 'CheckboxGroup',
  Checkbox = 'Checkbox',
  Radio = 'Radio',
  RadioGroup = 'RadioGroup',
  Select = 'Select',
  Text = 'Text',
}

export enum AddressField {
  Cc = 'cc',
  City = 'city',
  Number = 'number',
  PostalCode = 'postalCode',
  Street = 'street',
}

export enum OptionGroup {
  PackageType = 'packageType',
  Delivery = 'delivery',
  ShipmentOption = 'shipmentOption',
  Feature = 'feature',
  DropOff = 'dropOff',
}

export enum OptionType {
  String = 'string',
  Boolean = 'boolean',
  Number = 'number',
  Currency = 'currency',
  Date = 'date',
  Time = 'time',
  Select = 'select',
}

export const QUERY_KEY_CARRIERS = 'carriers';

export const QUERY_KEY_DELIVERY_OPTIONS = 'deliveryOptions';

export const QUERY_KEY_PICKUP_LOCATIONS = 'pickupLocations';
