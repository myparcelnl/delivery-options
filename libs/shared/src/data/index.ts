export * from './config';
export * from './constants';
export * from './enums';
export * from './errors';
export * from './getDefaultCarrierSettings';
export * from './getDefaultDeliveryOptionsConfig';
export * from './settingsConsts';
export * from './stringsKeys';
