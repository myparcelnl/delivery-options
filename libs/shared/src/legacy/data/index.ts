export * from './carrierFeatures';
export * from './carriers';
export * from './defaultAddress';
export * from './keys';
