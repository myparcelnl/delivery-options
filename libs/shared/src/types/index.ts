export * from './address.types';
export * from './carriers.types';
export * from './common.types';
export * from './config.types';
export * from './element.types';
export * from './form.types';
export * from './input.types';
export * from './options.types';
export * from './platform.types';
export * from './query.types';
export * from './validator.types';
