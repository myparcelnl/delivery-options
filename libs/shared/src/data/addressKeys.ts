/* eslint-disable no-magic-numbers */
import {AddressField} from '../types';

export const ADDRESS_CASE_MAP = {
  postal_code: AddressField.PostalCode,
};

export const ADDRESS_FIELD_COMBINATIONS = [
  [AddressField.PostalCode, AddressField.Number],
  [AddressField.PostalCode, AddressField.Street],
  [AddressField.PostalCode, AddressField.City],
  [AddressField.City, AddressField.Street],
];

export const ADDRESS_FIELDS = [
  {
    name: AddressField.Street,
    attributes: {
      autocomplete: 'address-line1',
    },
  },
  {
    name: AddressField.Number,
    attributes: {
      autocomplete: 'off',
      type: 'number',
      min: 1,
      max: 99999,
    },
  },
  {
    name: AddressField.PostalCode,
    attributes: {
      autocomplete: 'postal-code',
    },
  },
  {
    name: AddressField.City,
    attributes: {
      autocomplete: 'address-level2',
    },
  },
];
