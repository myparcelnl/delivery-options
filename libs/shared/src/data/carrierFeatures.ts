import {type ConfigKey} from '../types';
import {CarrierSetting, ConfigSetting} from '../enums';

type DeliveryOptionsFeature = readonly ConfigKey[];

export const FEATURES_DELIVERY: DeliveryOptionsFeature = Object.freeze([
  CarrierSetting.AllowDeliveryOptions,
  CarrierSetting.PriceStandardDelivery,
]);

export const FEATURES_EVENING_DELIVERY: DeliveryOptionsFeature = Object.freeze([
  CarrierSetting.AllowEveningDelivery,
  CarrierSetting.PriceEveningDelivery,
]);

export const FEATURES_MONDAY_DELIVERY: DeliveryOptionsFeature = Object.freeze([CarrierSetting.AllowMondayDelivery]);

export const FEATURES_MORNING_DELIVERY: DeliveryOptionsFeature = Object.freeze([
  CarrierSetting.AllowMorningDelivery,
  CarrierSetting.PriceMorningDelivery,
]);

export const FEATURES_ONLY_RECIPIENT: DeliveryOptionsFeature = Object.freeze([
  CarrierSetting.PriceOnlyRecipient,
  CarrierSetting.AllowOnlyRecipient,
]);

export const FEATURES_PACKAGE_TYPE_DIGITAL_STAMP: DeliveryOptionsFeature = Object.freeze([
  CarrierSetting.AllowPackageTypeDigitalStamp,
]);

export const FEATURES_PACKAGE_TYPE_MAILBOX: DeliveryOptionsFeature = Object.freeze([
  CarrierSetting.AllowPackageTypeMailbox,
]);

export const FEATURES_PICKUP: DeliveryOptionsFeature = Object.freeze([
  CarrierSetting.AllowPickupLocations,
  CarrierSetting.PricePickup,
]);

export const FEATURES_SATURDAY_DELIVERY: DeliveryOptionsFeature = Object.freeze([CarrierSetting.AllowSaturdayDelivery]);

export const FEATURES_SIGNATURE: DeliveryOptionsFeature = Object.freeze([
  CarrierSetting.AllowSignature,
  CarrierSetting.PriceSignature,
]);

export const FEATURES_SHOW_DELIVERY_DATE: DeliveryOptionsFeature = Object.freeze([ConfigSetting.ShowDeliveryDate]);

export const FEATURES_DROP_OFF_DAYS: DeliveryOptionsFeature = Object.freeze([CarrierSetting.DropOffDays]);

export const FEATURES_DROP_OFF_DELAY: DeliveryOptionsFeature = Object.freeze([CarrierSetting.DropOffDelay]);

export const FEATURES_DELIVERY_DAYS_WINDOW: DeliveryOptionsFeature = Object.freeze([CarrierSetting.DeliveryDaysWindow]);

export const FEATURES_CUTOFF_TIME: DeliveryOptionsFeature = Object.freeze([CarrierSetting.CutoffTime]);

export const FEATURES_SAME_DAY_DELIVERY: DeliveryOptionsFeature = Object.freeze([
  CarrierSetting.AllowSameDayDelivery,
  CarrierSetting.PriceSameDayDelivery,
  CarrierSetting.CutoffTimeSameDay,
]);

export const FEATURE_GROUPS = Object.freeze([
  FEATURES_DELIVERY,
  FEATURES_EVENING_DELIVERY,
  FEATURES_MONDAY_DELIVERY,
  FEATURES_MORNING_DELIVERY,
  FEATURES_ONLY_RECIPIENT,
  FEATURES_PACKAGE_TYPE_DIGITAL_STAMP,
  FEATURES_PACKAGE_TYPE_MAILBOX,
  FEATURES_PICKUP,
  FEATURES_SATURDAY_DELIVERY,
  FEATURES_SIGNATURE,
  FEATURES_SHOW_DELIVERY_DATE,
  FEATURES_DROP_OFF_DAYS,
  FEATURES_DROP_OFF_DELAY,
  FEATURES_DELIVERY_DAYS_WINDOW,
  FEATURES_CUTOFF_TIME,
  FEATURES_SAME_DAY_DELIVERY,
]) satisfies readonly DeliveryOptionsFeature[];
