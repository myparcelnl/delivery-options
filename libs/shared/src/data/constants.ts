import {DeliveryTypeName, PackageTypeName, PlatformName, ShipmentOptionName} from '@myparcel/constants';
import {type DropOffEntryObject} from '../types';
import {CarrierSetting} from './enums';

export const SUPPORTED_PLATFORMS = [PlatformName.MyParcel, PlatformName.SendMyParcel] as const;

export const SUPPORTED_DELIVERY_TYPES = [
  DeliveryTypeName.Standard,
  DeliveryTypeName.Evening,
  DeliveryTypeName.Morning,
  DeliveryTypeName.Pickup,
] as const;

export const SUPPORTED_SHIPMENT_OPTIONS = [ShipmentOptionName.OnlyRecipient, ShipmentOptionName.Signature] as const;

export const SUPPORTED_PACKAGE_TYPES = [
  PackageTypeName.Package,
  PackageTypeName.Mailbox,
  PackageTypeName.DigitalStamp,
  PackageTypeName.PackageSmall,
] as const;

export const DEFAULT_PLATFORM = PlatformName.MyParcel;

export const FORM_NAME_DELIVERY_OPTIONS = 'deliveryOptions';

export const REQUEST_KEY_CARRIERS = 'carriers';

export const REQUEST_KEY_DELIVERY_OPTIONS = 'deliveryOptions';

export const REQUEST_KEY_PICKUP_LOCATIONS = 'pickupLocations';

export const DROP_OFF_WEEKDAY = 'weekday' satisfies keyof DropOffEntryObject;

export const DROP_OFF_CUTOFF_TIME = CarrierSetting.CutoffTime satisfies keyof DropOffEntryObject;

export const DROP_OFF_SAME_DAY_CUTOFF_TIME = CarrierSetting.CutoffTimeSameDay satisfies keyof DropOffEntryObject;

export const DAY_MONDAY = 1;

export const DAY_TUESDAY = 2;

export const DAY_WEDNESDAY = 3;

export const DAY_THURSDAY = 4;

export const DAY_FRIDAY = 5;

export const DAY_SATURDAY = 6;

export const DAY_SUNDAY = 0;

export const DAYS_IN_WEEK = 7;

export const NBSP = '\u00A0';

export const CARRIER_IDENTIFIER_SEPARATOR = ':';

export const API_DATE_FORMAT = 'y-MM-dd kk:mm:ss';

export const DEFAULT_TIMEZONE_TYPE = 3;

export const DEFAULT_TIMEZONE = 'Europe/Amsterdam';

export const DELIVERY_TIMEFRAME_TYPE_START = 'start';

export const DELIVERY_TIMEFRAME_TYPE_END = 'end';
