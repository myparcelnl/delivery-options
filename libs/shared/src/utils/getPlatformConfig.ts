import {useMemoize} from '@vueuse/core';
import {configMyParcel, configSendMyParcel} from '@myparcel-do/shared';
import {PlatformName} from '@myparcel/constants';
import {type SupportedPlatformName} from '../types';

export const getPlatformConfig = useMemoize((platform: SupportedPlatformName) => {
  switch (platform) {
    case PlatformName.MyParcel:
      return configMyParcel();

    case PlatformName.SendMyParcel:
      return configSendMyParcel();
  }
});
