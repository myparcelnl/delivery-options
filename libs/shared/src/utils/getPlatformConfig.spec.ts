import {describe, expect, it} from 'vitest';
import {type SupportedPlatformName} from '../types';
import {SUPPORTED_PLATFORMS} from '../data/constants';
import {getPlatformConfig} from './getPlatformConfig';

describe('getPlatformConfig', () => {
  it.each(SUPPORTED_PLATFORMS)('gets config for platform %s', (platform) => {
    const config = getPlatformConfig(platform);

    expect(config).toMatchSnapshot();
  });

  it('throws error when no configuration is found', () => {
    expect(() => {
      return getPlatformConfig('foo' as SupportedPlatformName);
    }).toThrowError();
  });
});
