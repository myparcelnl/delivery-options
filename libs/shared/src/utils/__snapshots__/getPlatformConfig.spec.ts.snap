// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getPlatformConfig > gets config for platform belgie 1`] = `
{
  "carriers": [
    {
      "addressFields": [
        "postalCode",
        "street",
        "city",
      ],
      "deliveryCountries": [
        "BE",
        "NL",
      ],
      "deliveryTypes": [
        "standard",
        "pickup",
        "saturday",
      ],
      "features": [
        "deliveryDaysWindow",
        "dropOffDays",
        "dropOffDelay",
      ],
      "name": "bpost",
      "packageTypes": [
        "package",
        "mailbox",
      ],
      "pickupCountries": [
        "BE",
        "NL",
      ],
      "shipmentOptions": [
        "only_recipient",
        "signature",
      ],
      "subscription": 0,
    },
    {
      "deliveryCountries": [
        "BE",
        "NL",
      ],
      "deliveryTypes": [
        "standard",
        "pickup",
      ],
      "features": [
        "deliveryDaysWindow",
        "dropOffDays",
        "dropOffDelay",
      ],
      "name": "postnl",
      "packageTypes": [
        "package",
      ],
      "pickupCountries": [
        "BE",
        "NL",
      ],
      "shipmentOptions": [
        "signature",
      ],
      "subscription": 0,
    },
    {
      "addressFields": [
        "postalCode",
        "street",
        "city",
      ],
      "deliveryCountries": [
        "AT",
        "BE",
        "BG",
        "CZ",
        "DK",
        "EE",
        "FI",
        "FR",
        "DE",
        "GR",
        "HU",
        "IE",
        "IT",
        "LV",
        "LI",
        "LT",
        "LU",
        "NL",
        "PL",
        "PT",
        "RO",
        "SK",
        "SI",
        "ES",
        "SE",
      ],
      "deliveryTypes": [
        "standard",
        "pickup",
      ],
      "features": [
        "dropOffDays",
        "dropOffDelay",
      ],
      "name": "dpd",
      "packageTypes": [
        "package",
      ],
      "pickupCountries": [
        "AT",
        "BE",
        "CZ",
        "DK",
        "EE",
        "FI",
        "FR",
        "DE",
        "HU",
        "LV",
        "LT",
        "LU",
        "NL",
        "PL",
        "PT",
        "SK",
        "SI",
        "ES",
        "GB",
      ],
      "subscription": 0,
    },
  ],
}
`;

exports[`getPlatformConfig > gets config for platform myparcel 1`] = `
{
  "carriers": [
    {
      "addressFields": [
        "postalCode",
        "street",
        "city",
      ],
      "deliveryCountries": [
        "NL",
        "BE",
      ],
      "deliveryTypes": [
        "standard",
        "morning",
        "evening",
        "pickup",
        "monday",
      ],
      "fakeDelivery": true,
      "features": [
        "deliveryDaysWindow",
        "dropOffDays",
        "dropOffDelay",
      ],
      "name": "postnl",
      "packageTypes": [
        "package",
        "mailbox",
        "digital_stamp",
        "package_small",
      ],
      "pickupCountries": [
        "NL",
        "BE",
        "DK",
        "SE",
        "DE",
      ],
      "shipmentOptions": [
        "only_recipient",
        "signature",
      ],
      "subscription": -1,
    },
    {
      "addressFields": [
        "city",
        "postalCode",
      ],
      "deliveryCountries": [
        "NL",
        "BE",
      ],
      "deliveryTypes": [
        "standard",
        "pickup",
        "same_day",
      ],
      "features": [
        "deliveryDaysWindow",
        "dropOffDays",
        "dropOffDelay",
      ],
      "name": "dhlforyou",
      "packageTypes": [
        "package",
        "mailbox",
      ],
      "pickupCountries": [
        "NL",
      ],
      "shipmentOptions": [
        "only_recipient",
        "signature",
      ],
      "subscription": -1,
    },
    {
      "deliveryCountries": [
        "AT",
        "BG",
        "HR",
        "CZ",
        "EE",
        "FI",
        "DE",
        "GR",
        "HU",
        "IE",
        "IT",
        "LV",
        "LT",
        "LU",
        "PL",
        "PT",
        "RO",
        "SK",
        "SI",
        "ES",
      ],
      "deliveryTypes": [
        "standard",
        "pickup",
      ],
      "features": [],
      "name": "dhlparcelconnect",
      "packageTypes": [
        "package",
      ],
      "pickupCountries": [
        "AT",
        "BG",
        "HR",
        "CZ",
        "DK",
        "FR",
        "DE",
        "GR",
        "HU",
        "IT",
        "LU",
        "PL",
        "PT",
        "RO",
        "SK",
        "SI",
        "ES",
        "SE",
      ],
      "subscription": -1,
    },
    {
      "deliveryCountries": [
        "BE",
        "BG",
        "DK",
        "DE",
        "EE",
        "FI",
        "FR",
        "GR",
        "HU",
        "IE",
        "IT",
        "HR",
        "LV",
        "LT",
        "LU",
        "NL",
        "AT",
        "PL",
        "PT",
        "RO",
        "SI",
        "SK",
        "ES",
        "CZ",
        "SE",
        "GB",
      ],
      "deliveryTypes": [
        "standard",
      ],
      "features": [],
      "name": "dhleuroplus",
      "packageTypes": [
        "package",
      ],
      "subscription": -1,
    },
    {
      "addressFields": [
        "postalCode",
        "street",
        "city",
      ],
      "deliveryCountries": [
        "BG",
        "DE",
        "EE",
        "FI",
        "GR",
        "HU",
        "IE",
        "IT",
        "HR",
        "LV",
        "LT",
        "LU",
        "AT",
        "PL",
        "PT",
        "RO",
        "SI",
        "SK",
        "ES",
        "CZ",
      ],
      "deliveryTypes": [
        "standard",
        "pickup",
      ],
      "fakeDelivery": true,
      "fakeDeliveryBlacklist": [
        "NL",
      ],
      "name": "ups",
      "packageTypes": [
        "package",
      ],
      "pickupCountries": [
        "DE",
      ],
      "subscription": 1,
      "unsupportedParameters": [
        "package_type",
      ],
    },
    {
      "addressFields": [
        "postalCode",
        "street",
        "city",
      ],
      "deliveryCountries": [
        "AT",
        "BE",
        "BG",
        "CZ",
        "DK",
        "EE",
        "FI",
        "FR",
        "DE",
        "GR",
        "HU",
        "IE",
        "IT",
        "LV",
        "LI",
        "LT",
        "LU",
        "NL",
        "PL",
        "PT",
        "RO",
        "SK",
        "SI",
        "ES",
        "SE",
      ],
      "deliveryTypes": [
        "standard",
        "pickup",
      ],
      "features": [
        "dropOffDays",
        "dropOffDelay",
      ],
      "name": "dpd",
      "packageTypes": [
        "package",
        "mailbox",
      ],
      "pickupCountries": [
        "AT",
        "BE",
        "CZ",
        "DK",
        "EE",
        "FI",
        "FR",
        "DE",
        "HU",
        "LV",
        "LT",
        "LU",
        "NL",
        "PL",
        "PT",
        "SK",
        "SI",
        "ES",
        "GB",
      ],
      "subscription": 1,
    },
  ],
}
`;
